### DISCUSSION TITLE
Migration from J2V8

### DISCUSSION THREAD

**Discussion by newk5** [caoccao/Javet#9] [Q&A] [answered]
Hello, I recently just found out about this project and first of all I'd just like to say thanks for all the great work you've done and also for providing a more up to date version of J2V8, that was really useful :) 

I have an existent application that relies on J2V8 and I was thinking about migrating to Javet because it seems to be a much better alternative and actively maintained aswell. But I'm not sure Javet supports all the features I need so I want to ask first before I put in the effort of migrating the application. Some of the features I already noticed are present but I'm not sure about some of them. Here's the features my application is currently using with J2V8:

1. Calling a Javascript function from Java with arguments. For example if I have a method:

```java
public void doThing(MyCustomObject obj) {
     v8Context.executeJSFunction("doThing", obj); //being able to pass MyCustomObject to the javascript context
}
```

2. Calling a Java function from Javascript, passing several javascript arguments and being able to access them in Java
3. Being able to "inject" custom java objects into the V8 context (In J2V8 I used a 3rd party library to do this https://github.com/alicorn-systems/v8-adapter which worked quite well). 

Can you confirm if all those 3 features are possible with Javet at the moment?

**Comment by caoccao** [comment, marked as answer]
Thank you for asking these questions.

Indeed, you are not alone because I faced these questions before creating Javet. Short answer is yes, all these 3 features are supported.

I created a few tutorials that cover all of them. I suggest you spend some time evaluating the tutorials and sample code.

I migrated my personal projects to Javet from v0.7.0. The migration was a little bit painful, but very exciting and there was no outstanding issues. It just worked as expected. Good luck to you. Of course, if you face any issues, I will be very happy to help investigate or fix them.

**Reply by newk5** [reply]
Thanks, I'm gonna look into it.

**Comment by newk5** [comment]
@caoccao I looked into the documentation but I'm  not being able to achieve 1. and 3. Maybe I'm not doing it right, here's my code I wrote to test it:
```java
       //create custom object
        TestObject test = new TestObject();
        test.setName("abc");
        test.setType("test");

        //convert it to a V8Value
        JavetObjectConverter converter = new JavetObjectConverter();
        V8Runtime v8 = V8Host.getNodeInstance().createV8Runtime();
        V8Value val = converter.toV8Value(v8, test);

        //load the script
        V8Runtime script = v8.getExecutor(new File("C:\\main.js")).getV8Runtime();

        //call javascript function: helloFunc(value) and pass the test object
        script.invoke(script.getGlobalObject(), "helloFunc", false, val);
```
My main.js:
```javascript
function helloFunc(testValue){
    console.log(testValue.getName());
}

```

I tried it with a node instance and a normal instance. The converter seems to convert my test object to V8Undefined, but even if I pass a string instead of my custom object, the javascript function is never triggered.

**Reply by caoccao** [reply]
Goals
---

I see what you were trying to achieve. Let me separate them into 2 goals.

1. **A generic POJO converter** - That can be easily done by a few lines of code. I didn't build it into `JavetObjectConverter` because I wanted `JavetObjectConverter` to (a) cover primitive types + Set + Map only; (b) completely open to subclass; (c) not bring performance overhead. So in your case, a POJO converter has to deal with reflection which is so slow that I decided to leave it to applications.
2. **Thinking in Javet** - I liked the major API exposed by J2V8 but disliked the others. I invented the Javet flavored API. Of cource, that requires a little bit mind shift or **Thinking in Javet**.

Let's do it in the Javet way
---

Let's say you have a Pojo that allows you to define a name-value pair.

``` java
package com.caoccao.javet.thinkinginjavet.pojo;

public class Pojo {
    private String name;
    private String value;

    public Pojo() {
        this(null, null);
    }

    public Pojo(String name, String value) {
        this.name = name;
        this.value = value;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getValue() {
        return value;
    }

    public void setValue(String value) {
        this.value = value;
    }
}
```

PojoConverter
---

Then, you need a generic PojoConverter.

* It is generic enough to cover all kinds of Pojo objects in a recursive way.
* There is no need to deal with primitive types because the parent converter handles that.

``` java
package com.caoccao.javet.thinkinginjavet.pojo;

import com.caoccao.javet.exceptions.JavetException;
import com.caoccao.javet.interop.V8Runtime;
import com.caoccao.javet.utils.converters.JavetObjectConverter;
import com.caoccao.javet.values.V8Value;
import com.caoccao.javet.values.primitive.V8ValueString;
import com.caoccao.javet.values.reference.V8ValueObject;

import java.lang.reflect.Method;
import java.util.HashSet;
import java.util.Locale;
import java.util.Set;

@SuppressWarnings("unchecked")
public class PojoConverter extends JavetObjectConverter {
    public static final String METHOD_PREFIX_GET = "get";
    public static final String METHOD_PREFIX_IS = "is";
    protected static final Set<String> EXCLUDED_METHODS;

    static {
        EXCLUDED_METHODS = new HashSet<>();
        for (Method method : Object.class.getMethods()) {
            if (method.getParameterCount() == 0) {
                String methodName = method.getName();
                if (methodName.startsWith(METHOD_PREFIX_IS) || methodName.startsWith(METHOD_PREFIX_GET)) {
                    EXCLUDED_METHODS.add(methodName);
                }
            }
        }
    }

    @Override
    public V8Value toV8Value(V8Runtime v8Runtime, Object object) throws JavetException {
        V8Value v8Value = super.toV8Value(v8Runtime, object);
        if (v8Value != null && !(v8Value.isUndefined())) {
            return v8Value;
        }
        Class objectClass = object.getClass();
        V8ValueObject v8ValueObject = v8Runtime.createV8ValueObject();
        for (Method method : objectClass.getMethods()) {
            if (method.getParameterCount() == 0 && method.canAccess(object)) {
                String methodName = method.getName();
                String propertyName = null;
                if (methodName.startsWith(METHOD_PREFIX_IS) && !EXCLUDED_METHODS.contains(methodName)
                        && methodName.length() > METHOD_PREFIX_IS.length()) {
                    propertyName = methodName.substring(METHOD_PREFIX_IS.length(), METHOD_PREFIX_IS.length() + 1).toLowerCase(Locale.ROOT)
                            + methodName.substring(METHOD_PREFIX_IS.length() + 1);
                } else if (methodName.startsWith(METHOD_PREFIX_GET) && !EXCLUDED_METHODS.contains(methodName)
                        && methodName.length() > METHOD_PREFIX_GET.length()) {
                    propertyName = methodName.substring(METHOD_PREFIX_GET.length(), METHOD_PREFIX_GET.length() + 1).toLowerCase(Locale.ROOT)
                            + methodName.substring(METHOD_PREFIX_GET.length() + 1);
                }
                if (propertyName != null) {
                    try (V8Value v8ValueTemp = toV8Value(v8Runtime, method.invoke(object))) {
                        v8ValueObject.set(propertyName, v8ValueTemp);
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            }
        }
        v8Value = v8ValueObject;
        return v8Runtime.decorateV8Value(v8Value);
    }
}
```

Let's make use of it!
---

* There are only few lines actually involves Javet API to achieve your goals.
* Context is a relative concept. In this case, it may be global object which is `globalThis`. The sample code set the pojo object to `globalThis`.
* `console.log()` is able to see the pojo object.

``` java
package com.caoccao.javet.thinkinginjavet.pojo;

import com.caoccao.javet.exceptions.JavetException;
import com.caoccao.javet.interop.V8Host;
import com.caoccao.javet.interop.V8Runtime;
import com.caoccao.javet.values.V8Value;

public class TestPojo {
    public static void main(String[] args) throws JavetException {
        Pojo[] pojoArray = new Pojo[]{
                new Pojo("Tom", "CEO"),
                new Pojo("Jerry", "CFO")};
        PojoConverter converter = new PojoConverter();
        try (V8Runtime v8Runtime = V8Host.getNodeInstance().createV8Runtime()) {
            try (V8Value pojoContext = converter.toV8Value(v8Runtime, pojoArray)) {
                v8Runtime.getGlobalObject().set("pojoContext", pojoContext);
            }
            v8Runtime.getExecutor("console.log(pojoContext);").executeVoid();
        }
    }
}
```

The console output is:

``` json
[ { name: 'Tom', value: 'CEO' }, { name: 'Jerry', value: 'CFO' } ]
```

Hope that answers your question. The mind shift may bring you out of the comfort zone created by J2V8. Please stay calm and you will be happy soon.

Please let me know if you have any other questions. Good luck with your migration.

**Reply by newk5** [reply]
Thanks! That helps alot with the POJO to JS Object conversion. But how about invoking a function on my .js file and passing arguments to it? Inside my main.js file I will have many functions, and the idea is to only read that file once (like on my example above) and then simply call the functions inside it when needed and pass the java values (converted to V8Values using POJO converter). My use case is a game server that receives events in Java, and the idea was to pass those events to the V8 context with their arguments. In J2V8 I could just read my main.js file once (v8.executeScript("main.js") ) and then just call v8.executeJSFunction("helloFun", arg1, arg2, etc...) on the several java events that were triggered. In Javet, I dont know how I would do this. I tried to replicate in my example above. Read main.js once and then just call the functions inside it when needed.

Btw have you considered creating a gitter chatroom for Javet? It might be easier to communicate that way 😛

**Reply by caoccao** [reply]
I wonder if this [tutorial](https://github.com/caoccao/Javet/blob/main/docs/tutorial/manipulate_v8_function.rst) is what you want.

As of **Thinking in Javet**, `V8Runtime` is no longer responsible for calling functions. It's within caller's scope. So, in your case, `v8Runtime.getGlobalObject().invoke(functionName, param1, param2, ...)`.

Regarding the chatroom, I'm completely open to that idea. Any suggested one?

**Reply by newk5** [reply]
I tried globalObject.invoke() but also didn't work, I must be doing something wrong hmm...

Check out https://gitter.im/ you can create chatroom and link it to this github repo

**Reply by newk5** [reply]
Problem solved! Thanks, I forgot to call excuteVoid() on the executor
